cookbook/core/control_flow/dynamics.py

```diff
@@ -1,56 +1,122 @@
+        return ord(character) - ord("a")
+    else:
+        return ord(character) - ord("A")
 
+
+# %%
+# We now write a task that prepares the 26-character list by populating the frequency of every character.
 @task
-def t1(a: int) -> str:
-    a = a + 2
-    return "world-" + str(a)
+def update_list(freq_list: typing.List[int], list_index: int) -> typing.List[int]:
+    """
+    Notes the frequency of characters"""
+    freq_list[list_index] += 1
+    return freq_list
 
 
+# %%
+# Next we find the number of common characters between the two strings.
 @task
-def t2(a: str, b: str) -> str:
-    return b + a
+def derive_count(freq1: typing.List[int], freq2: typing.List[int]) -> int:
+    """
+    Derives the number of common characters"""
+    count = 0
+    for i in range(26):
+        count += min(freq1[i], freq2[i])
+    return count
 
 
 # %%
-# This is how you declare a dynamic task. Note that you can call ``range`` on the input ``a``. In a normal workflow,
-# you cannot do this and compilation would fail very early on. Because this isn't run until there are literal values
-# for all inputs, it does work.
+# In this step, we perform the following:
+#
+# #. Initialize the empty 26-character list to be sent to the ``update_list`` task
+# #. Loop through every character of the first string (s1) and populate the frequency list
+# #. Loop through every character of the second string (s2) and populate the frequency list
+# #. Derive the number of common characters by comparing the two frequency lists
+#
+# The looping is dependent on the number of characters of both the strings which aren't known until the run time. If the ``@task`` decorator is used to encapsulate the calls mentioned above, the compilation will fail very early on due to the absence of the literal values.
+# Therefore, ``@dynamic`` decorator has to be used.
+#
+# Dynamic workflow is effectively both a task and a workflow. The key thing to note is that the ``body of tasks is run at run time and the
+# body of workflows is run at compile (aka registration) time``. Essentially, this is what a dynamic workflow leverages -- it’s a workflow that is compiled at run time (the best of both worlds)!
 #
-# Local execution will also work as it's treated like a task that will be run with Python native inputs.
+# At execution (run) time, Flytekit runs the compilation step, and produces
+# a ``WorkflowTemplate`` (from the dynamic workflow), which Flytekit then passes back to Flyte Propeller for further running, much like how sub-workflows are handled.
 #
-# The name of this function was intentionally chosen. Even though these are implemented as tasks in flytekit, you
-# should think about them as workflows. At execution time, flytekit is running the compilation step, and producing
-# a ``WorkflowTemplate``, which it then passes back to Flyte Propeller for further running, much like how sub-workflows
-# are handled.
+# We now define the dynamic workflow encapsulating the above mentioned points.
```

Comment:
I think we should add a note linking to map tasks at the very least, explaining that those are more efficient in certain cases.  I just don't want people to default to the dynamic pattern for iterating over a list since it's not the most efficient.  Map tasks only work for Python tasks (tasks decorated with the `@task` decorator, not sql/spark/etc) so we should note that as well.

https://github.com/flyteorg/flytekit/blob/8528268a29a07fe7e9ce9f7f08fea68c41b6a60b/flytekit/core/map_task.py

Reply:
Please check if the note was added correctly. Thanks.

Reply:
Can we chat some time about cross linking?

Reply:
Sure!